airflow-core/src/airflow/migrations/versions/0072_3_1_0_modify_taskinstance_dag_version_fk.py

```diff
@@ -0,0 +1,72 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""
+Modify task_instance.dag_version_id FK on delete behavior to SET NULL.
+
+Revision ID: 1b2cd3e4f5a6
+Revises: 03e36c7f30aa
+Create Date: 2025-05-21 00:00:00.000000
+"""
+
+from __future__ import annotations
+
+from alembic import op
+from sqlalchemy_utils import UUIDType
+
+# revision identifiers, used by Alembic.
+revision = "1b2cd3e4f5a6"
+down_revision = "03e36c7f30aa"
+branch_labels = None
+depends_on = None
+airflow_version = "3.1.0"
+
+
+def upgrade():
+    """Alter task_instance.dag_version_id to SET NULL on delete."""
+    with op.batch_alter_table("task_instance", schema=None) as batch_op:
+        batch_op.drop_constraint(batch_op.f("task_instance_dag_version_id_fkey"), type_="foreignkey")
+        batch_op.alter_column(
+            "dag_version_id",
+            existing_type=UUIDType(binary=False),
+            nullable=True,
+        )
+        batch_op.create_foreign_key(
+            batch_op.f("task_instance_dag_version_id_fkey"),
+            "dag_version",
+            ["dag_version_id"],
+            ["id"],
+            ondelete="SET NULL",
```

Comment:
Hmm, do we need `ondelete=set null"`? If DagVersion is soft deleted, we wouldn’t delete any DagVersion from the database, so these foreign keys can be kept without being set to null, right?

Reply:
Just want to make sure I understand correctly. when we fetch dag_version from task_instance , we 'll need to change it to fetch those with `is_active=True`, right?

If that's the case, do we need to update the existing DagVersion.is_active to True?

Reply:
> Hmm, do we need `ondelete=set null"`? If DagVersion is soft deleted, we wouldn’t delete any DagVersion from the database, so these foreign keys can be kept without being set to null, right?

Right we don't need `ondelete=set null` as it is a soft delete.

Reply:
> Just want to make sure I understand correctly. when we fetch dag_version from task_instance , we 'll need to change it to fetch those with `is_active=True`, right?
> 
> If that's the case, do we need to update the existing DagVersion.is_active to True?

No, we don’t need to back-fill or update existing rows – that was done in 0071 migration. It has a `server_default=1` (so all _new_ rows are active by default).

In airflow-core/src/airflow/models/dag_version.py, as part of this commit, in DagVersion helpers (`DagVersion.get_latest_version()` and `DagVersion.get_version()` ) already include `WHERE is_active = true` in their SELECTs.

Do we want `task_instance.dag_version` to only ever return active versions? I don't think so, it should return all task instances? Isn't it?